File: ruby/test/ws_transport_test.rb
# frozen_string_literal: true

require_relative "test_helper"

# Unit tests for DatagroutConduit::Transport::Ws
#
# These tests inject frames directly into the transport's handle_message method
# (via #send) and inspect state via instance_variable_get, bypassing the real
# WebSocket socket so no live server is needed.

class WsTransportTest < Minitest::Test
  # ── Helpers ───────────────────────────────────────────────────────────────

  # Build a Ws instance with @connected = true and a fake driver
  # (no real socket, no real WS handshake).
  def make_ws
    ws = DatagroutConduit::Transport::Ws.new(
      url: "wss://example.com/ws",
      auth: { bearer: "test-token" }
    )
    ws.instance_variable_set(:@connected, true)
    ws.instance_variable_set(:@next_id, 0)
    ws
  end

  # Directly invoke the private handle_message method.
  def inject(ws, payload)
    ws.send(:handle_message, JSON.generate(payload))
  end

  # Retrieve the private @pending map.
  def pending(ws)
    ws.instance_variable_get(:@pending)
  end

  def pending_subscribe(ws)
    ws.instance_variable_get(:@pending_subscribe)
  end

  def subscriptions(ws)
    ws.instance_variable_get(:@subscriptions)
  end

  # ── URL rewriting ─────────────────────────────────────────────────────────

  def test_url_rewrite_https_to_wss
    client = DatagroutConduit::Client.new(
      url: "https://gateway.datagrout.ai/servers/test/ws",
      auth: { bearer: "tok" },
      transport: :websocket
    )
    assert_equal "wss://gateway.datagrout.ai/servers/test/ws",
                 client.transport.instance_variable_get(:@url)
  end

  def test_url_rewrite_http_to_ws
    client = DatagroutConduit::Client.new(
      url: "http://localhost:4000/ws",
      auth: { bearer: "tok" },
      transport: :websocket
    )
    assert_equal "ws://localhost:4000/ws",
                 client.transport.instance_variable_get(:@url)
  end

  def test_url_already_wss_unchanged
    client = DatagroutConduit::Client.new(
      url: "wss://gateway.datagrout.ai/servers/test/ws",
      auth: { bearer: "tok" },
      transport: :websocket
    )
    assert_equal "wss://gateway.datagrout.ai/servers/test/ws",
                 client.transport.instance_variable_get(:@url)
  end

  # ── Connection guard ──────────────────────────────────────────────────────

  def test_send_request_raises_when_not_connected
    ws = DatagroutConduit::Transport::Ws.new(url: "wss://example.com/ws", auth: {})
    assert_raises(DatagroutConduit::NotInitializedError) { ws.send_request("tools/list") }
  end

  def test_subscribe_raises_when_not_connected
    ws = DatagroutConduit::Transport::Ws.new(url: "wss://example.com/ws", auth: {})
    assert_raises(DatagroutConduit::NotInitializedError) { ws.subscribe("agents.x.events") }
  end

  # ── Response routing — regular requests ──────────────────────────────────

  def test_routes_response_to_pending_caller
    ws = make_ws

    future = DatagroutConduit::Transport::Ws::RequestFuture.new
    ws.instance_variable_get(:@mutex).synchronize do
      pending(ws)["ws-1"] = future
    end

    inject(ws, { "jsonrpc" => "2.0", "id" => "ws-1", "result" => { "tools" => [] } })

    result, value = future.wait(timeout: 1)
    assert_equal :ok, result
    assert_equal({ "tools" => [] }, value)
  end

  def test_routes_rpc_error_to_pending_caller
    ws = make_ws

    future = DatagroutConduit::Transport::Ws::RequestFuture.new
    ws.instance_variable_get(:@mutex).synchronize do
      pending(ws)["ws-1"] = future
    end

    inject(ws, {
      "jsonrpc" => "2.0",
      "id"      => "ws-1",
      "error"   => { "code" => -32_601, "message" => "Method not found" }
    })

    result, value = future.wait(timeout: 1)
    assert_equal :error, result
    assert_includes value, "Method not found"
  end

  def test_clears_pending_after_response
    ws = make_ws

    future = DatagroutConduit::Transport::Ws::RequestFuture.new
    ws.instance_variable_get(:@mutex).synchronize do
      pending(ws)["ws-2"] = future
    end

    inject(ws, { "jsonrpc" => "2.0", "id" => "ws-2", "result" => {} })
    future.wait(timeout: 1)

    assert_empty pending(ws)
  end

  def test_ignores_response_for_unknown_id
    ws = make_ws
    inject(ws, { "jsonrpc" => "2.0", "id" => "ghost", "result" => {} })
    assert_empty pending(ws)  # no crash, no state pollution
  end

  def test_ignores_malformed_json
    ws = make_ws
    ws.send(:handle_message, "not valid json{{{")
    assert_empty pending(ws)  # still alive, no error
  end

  def test_id_coercion_integer_ids
    ws = make_ws

    future = DatagroutConduit::Transport::Ws::RequestFuture.new
    ws.instance_variable_get(:@mutex).synchronize do
      pending(ws)["42"] = future
    end

    # Server sends integer id — should still match string key "42"
    inject(ws, { "jsonrpc" => "2.0", "id" => 42, "result" => { "ok" => true } })

    result, _value = future.wait(timeout: 1)
    assert_equal :ok, result
  end

  # ── Subscribe response routing ────────────────────────────────────────────

  def test_subscribe_response_registers_subscription
    ws = make_ws

    future = DatagroutConduit::Transport::Ws::RequestFuture.new
    ws.instance_variable_get(:@mutex).synchronize do
      pending_subscribe(ws)["req-1"] = { topic: "agents.x.events", future: future }
    end

    inject(ws, {
      "jsonrpc" => "2.0",
      "id"      => "req-1",
      "result"  => { "subscription" => "sub_abc", "topic" => "agents.x.events" }
    })

    result, value = future.wait(timeout: 1)
    assert_equal :ok, result
    assert_equal({ "subscription" => "sub_abc", "topic" => "agents.x.events" }, value)

    assert_empty pending_subscribe(ws)
  end

  def test_subscribe_error_response_rejects_future
    ws = make_ws

    future = DatagroutConduit::Transport::Ws::RequestFuture.new
    ws.instance_variable_get(:@mutex).synchronize do
      pending_subscribe(ws)["req-1"] = { topic: "agents.x.events", future: future }
    end

    inject(ws, {
      "jsonrpc" => "2.0",
      "id"      => "req-1",
      "error"   => { "code" => -32_000, "message" => "Topic not found" }
    })

    result, value = future.wait(timeout: 1)
    assert_equal :error, result
    assert_includes value, "Topic not found"
  end

  # ── Notification routing ──────────────────────────────────────────────────

  def test_delivers_notification_to_subscription
    ws = make_ws
    sub = DatagroutConduit::Transport::Ws::Subscription.new("sub_abc", "agents.x.events")
    ws.instance_variable_get(:@mutex).synchronize do
      subscriptions(ws)["sub_abc"] = [sub]
    end

    inject(ws, {
      "jsonrpc" => "2.0",
      "method"  => "notification",
      "params"  => {
        "subscription" => "sub_abc",
        "event"        => "agent.thought",
        "data"         => { "text" => "thinking" }
      }
    })

    event = sub.recv(timeout: 1)
    refute_nil event
    assert_equal "sub_abc",       event.subscription
    assert_equal "agent.thought", event.event
    assert_equal({ "text" => "thinking" }, event.data)
  end

  def test_silently_drops_notification_for_unknown_subscription
    ws = make_ws
    inject(ws, {
      "jsonrpc" => "2.0",
      "method"  => "notification",
      "params"  => { "subscription" => "ghost", "event" => "x", "data" => nil }
    })
    assert_empty subscriptions(ws)
  end

  def test_ignores_notification_without_subscription_field
    ws = make_ws
    inject(ws, {
      "jsonrpc" => "2.0",
      "method"  => "notification",
      "params"  => { "event" => "x", "data" => nil }
    })
    assert_empty subscriptions(ws)  # no crash
  end

  def test_ignores_unknown_server_methods
    ws = make_ws
    inject(ws, {
      "jsonrpc" => "2.0",
      "method"  => "session.ready",
      "params"  => { "session_id" => "abc" }
    })
    assert_empty subscriptions(ws)
  end

  # ── Disconnect ────────────────────────────────────────────────────────────

  def test_disconnect_clears_pending_and_subscriptions
    ws = make_ws

    future = DatagroutConduit::Transport::Ws::RequestFuture.new
    sub    = DatagroutConduit::Transport::Ws::Subscription.new("sub_1", "test.topic")

    ws.instance_variable_get(:@mutex).synchronize do
      pending(ws)["ws-1"] = future
      subscriptions(ws)["sub_1"] = [sub]
    end

    ws.send(:fail_all_pending, :disconnected)

    result, _ = future.wait(timeout: 1)
    assert_equal :error, result
    assert_predicate sub, :closed?
  end

  def test_fail_all_pending_closes_subscriptions
    ws = make_ws

    sub1 = DatagroutConduit::Transport::Ws::Subscription.new("s1", "t1")
    sub2 = DatagroutConduit::Transport::Ws::Subscription.new("s2", "t2")

    ws.instance_variable_get(:@mutex).synchronize do
      subscriptions(ws)["s1"] = [sub1]
      subscriptions(ws)["s2"] = [sub2]
    end

    ws.send(:fail_all_pending, :disconnected)

    assert_predicate sub1, :closed?
    assert_predicate sub2, :closed?
  end

  # ── Subscription class ────────────────────────────────────────────────────

  def test_subscription_recv_returns_event
    sub   = DatagroutConduit::Transport::Ws::Subscription.new("s1", "topic")
    event = DatagroutConduit::Transport::Ws::SubscriptionEvent.new(
      subscription: "s1", event: "test", data: { "x" => 1 }
    )

    sub._enqueue(event)
    assert_equal event, sub.recv(timeout: 1)
  end

  def test_subscription_recv_times_out
    sub = DatagroutConduit::Transport::Ws::Subscription.new("s1", "topic")
    result = sub.recv(timeout: 0.05)
    assert_nil result
  end

  def test_subscription_closed_raises_stop_iteration_on_recv
    sub = DatagroutConduit::Transport::Ws::Subscription.new("s1", "topic")
    sub._close

    assert_raises(StopIteration) { sub.recv }
  end

  def test_subscription_close_is_idempotent
    sub = DatagroutConduit::Transport::Ws::Subscription.new("s1", "topic")
    sub._close
    sub._close  # should not raise or double-enqueue sentinel
    assert_predicate sub, :closed?
  end

  def test_subscription_enqueue_ignores_after_close
    sub = DatagroutConduit::Transport::Ws::Subscription.new("s1", "topic")
    sub._close
    sub._enqueue(DatagroutConduit::Transport::Ws::SubscriptionEvent.new(
      subscription: "s1", event: "x", data: nil
    ))
    # Queue has only the nil sentinel; recv raises StopIteration
    assert_raises(StopIteration) { sub.recv }
  end

  def test_subscription_each_iterates_until_closed
    sub    = DatagroutConduit::Transport::Ws::Subscription.new("s1", "topic")
    events = []
    e1 = DatagroutConduit::Transport::Ws::SubscriptionEvent.new(
      subscription: "s1", event: "a", data: nil
    )
    e2 = DatagroutConduit::Transport::Ws::SubscriptionEvent.new(
      subscription: "s1", event: "b", data: nil
    )

    sub._enqueue(e1)
    sub._enqueue(e2)
    sub._close

    sub.each { |e| events << e }
    assert_equal [e1, e2], events
  end

  # ── Client integration ────────────────────────────────────────────────────

  def test_subscribe_raises_when_not_websocket_transport
    client = DatagroutConduit::Client.new(
      url: "https://example.com/mcp",
      auth: { bearer: "tok" },
      transport: :mcp
    )
    err = assert_raises(RuntimeError) { client.subscribe("agents.x.events") }
    assert_includes err.message, "websocket"
  end

  def test_unsubscribe_raises_when_not_websocket_transport
    client = DatagroutConduit::Client.new(
      url: "https://example.com/mcp",
      auth: { bearer: "tok" },
      transport: :mcp
    )
    err = assert_raises(RuntimeError) { client.unsubscribe("sub_123") }
    assert_includes err.message, "websocket"
  end

  def test_client_websocket_transport_uses_ws_class
    client = DatagroutConduit::Client.new(
      url: "wss://example.com/ws",
      auth: { bearer: "tok" },
      transport: :websocket
    )
    assert_instance_of DatagroutConduit::Transport::Ws, client.transport
  end

  def test_client_websocket_transport_string_key
    client = DatagroutConduit::Client.new(
      url: "wss://example.com/ws",
      auth: { bearer: "tok" },
      transport: "websocket"
    )
    assert_instance_of DatagroutConduit::Transport::Ws, client.transport
  end

  def test_unknown_transport_raises
    assert_raises(DatagroutConduit::ConfigError) do
      DatagroutConduit::Client.new(
        url: "https://example.com/mcp",
        auth: { bearer: "tok" },
        transport: :grpc
      )
    end
  end

  # ── Auth header building ──────────────────────────────────────────────────

  def test_bearer_auth_header
    ws = DatagroutConduit::Transport::Ws.new(
      url: "wss://example.com/ws",
      auth: { bearer: "my-token" }
    )
    headers = ws.send(:build_upgrade_headers)
    assert_equal "Bearer my-token", headers["Authorization"]
  end

  def test_api_key_auth_header
    ws = DatagroutConduit::Transport::Ws.new(
      url: "wss://example.com/ws",
      auth: { api_key: "ak-123" }
    )
    headers = ws.send(:build_upgrade_headers)
    assert_equal "ak-123", headers["X-API-Key"]
  end

  def test_basic_auth_header
    ws = DatagroutConduit::Transport::Ws.new(
      url: "wss://example.com/ws",
      auth: { basic: { username: "user", password: "pass" } }
    )
    headers = ws.send(:build_upgrade_headers)
    expected = "Basic #{Base64.strict_encode64("user:pass")}"
    assert_equal expected, headers["Authorization"]
  end

  def test_no_auth_produces_empty_headers
    ws = DatagroutConduit::Transport::Ws.new(url: "wss://example.com/ws", auth: {})
    headers = ws.send(:build_upgrade_headers)
    assert_empty headers
  end

  # ── Ping keepalive ──────────────────────────────────────────────────────────

  def test_ping_interval_constant_is_25_seconds
    # Mirror of Rust PING_INTERVAL; must stay aligned for parity.
    assert_equal 25, DatagroutConduit::Transport::Ws::PING_INTERVAL_SECONDS
  end

  def test_default_ping_interval_matches_constant
    ws = DatagroutConduit::Transport::Ws.new(url: "wss://example.com/ws")
    assert_in_delta DatagroutConduit::Transport::Ws::PING_INTERVAL_SECONDS, ws.ping_interval, 0.001
  end

  def test_constructor_accepts_ping_interval_override
    ws = DatagroutConduit::Transport::Ws.new(url: "wss://example.com/ws", ping_interval: 0.05)
    assert_in_delta 0.05, ws.ping_interval, 0.001
  end

  def test_pings_sent_starts_at_zero
    ws = DatagroutConduit::Transport::Ws.new(url: "wss://example.com/ws", ping_interval: 0.05)
    assert_equal 0, ws.pings_sent
  end

  def test_ping_thread_fires_pings_against_a_driver_double
    ws = DatagroutConduit::Transport::Ws.new(url: "wss://example.com/ws", ping_interval: 0.02)

    # Drive the ping loop without a real socket: install a driver double and
    # mark @connected so start_ping_thread keeps iterating.
    ping_calls = 0
    driver = Object.new
    driver.define_singleton_method(:ping) { ping_calls += 1; true }
    ws.instance_variable_set(:@driver, driver)
    ws.instance_variable_set(:@connected, true)

    ws.send(:start_ping_thread)
    sleep 0.08
    ws.send(:stop_ping_thread)
    ws.instance_variable_set(:@connected, false)

    assert ping_calls >= 1, "expected at least 1 ping in 80ms with 20ms interval; got #{ping_calls}"
    assert ws.pings_sent >= 1
  end

  def test_ping_thread_exits_when_driver_returns_false
    ws = DatagroutConduit::Transport::Ws.new(url: "wss://example.com/ws", ping_interval: 0.02)
    driver = Object.new
    driver.define_singleton_method(:ping) { false }  # simulates closing connection
    ws.instance_variable_set(:@driver, driver)
    ws.instance_variable_set(:@connected, true)

    ws.send(:start_ping_thread)
    sleep 0.08
    refute ws.instance_variable_get(:@ping_thread)&.alive?,
           "ping thread should exit when driver#ping returns false"
    ws.instance_variable_set(:@connected, false)
  end

  def test_stop_ping_thread_is_idempotent
    ws = DatagroutConduit::Transport::Ws.new(url: "wss://example.com/ws", ping_interval: 0.02)
    ws.send(:stop_ping_thread)  # nothing to stop
    ws.send(:stop_ping_thread)  # still nothing to stop
    assert true # if we got here without an exception, it's idempotent
  end
end

File: elixir/test/ws_transport_test.exs
defmodule DatagroutConduit.Transport.WsTest do
  @moduledoc """
  Unit tests for the WebSocket transport (datagrout-jsonrpc.v1).

  These tests exercise the GenServer message-routing logic by injecting
  raw frames directly into the process mailbox (simulating what the
  WebSockex connection module would deliver), without requiring a live
  WebSocket server.
  """

  use ExUnit.Case, async: true

  alias DatagroutConduit.Transport.Ws

  # ── Helpers ─────────────────────────────────────────────────────────────────

  defp inject_frame(pid, payload) do
    send(pid, {:ws_frame, Jason.encode!(payload)})
  end

  # ── Route regular RPC response ───────────────────────────────────────────────

  describe "handle_info/2 — regular response routing" do
    test "routes a JSON-RPC response to the pending caller" do
      {:ok, pid} =
        GenServer.start_link(Ws, %Ws{
          conn_pid: self(),
          pending: %{},
          pending_subscribe: %{},
          subscriptions: %{},
          next_id: 0
        })

      # Plant a pending request manually.
      :sys.replace_state(pid, fn state ->
        %{state | pending: Map.put(state.pending, "ws-1", {self(), make_ref()})}
      end)

      inject_frame(pid, %{"jsonrpc" => "2.0", "id" => "ws-1", "result" => %{"ok" => true}})

      # The GenServer calls GenServer.reply — but since we manually injected the
      # pending tuple, the reply goes nowhere. We verify state cleanup instead.
      Process.sleep(50)

      state = :sys.get_state(pid)
      assert map_size(state.pending) == 0

      GenServer.stop(pid)
    end

    test "ignores responses for unknown ids" do
      {:ok, pid} =
        GenServer.start_link(Ws, %Ws{
          conn_pid: self(),
          pending: %{},
          pending_subscribe: %{},
          subscriptions: %{},
          next_id: 0
        })

      inject_frame(pid, %{"jsonrpc" => "2.0", "id" => "ghost", "result" => %{}})
      Process.sleep(50)

      state = :sys.get_state(pid)
      assert map_size(state.pending) == 0

      GenServer.stop(pid)
    end

    test "ignores malformed JSON (not delivered as parsed map)" do
      {:ok, pid} =
        GenServer.start_link(Ws, %Ws{
          conn_pid: self(),
          pending: %{},
          pending_subscribe: %{},
          subscriptions: %{},
          next_id: 0
        })

      # Inject a non-JSON frame directly.
      send(pid, {:ws_frame, "not valid json"})
      Process.sleep(50)

      # GenServer should still be alive.
      assert Process.alive?(pid)

      GenServer.stop(pid)
    end
  end

  # ── Subscribe response routing ────────────────────────────────────────────────

  describe "handle_info/2 — subscribe response routing" do
    test "registers subscription and notifies subscriber on success" do
      {:ok, pid} =
        GenServer.start_link(Ws, %Ws{
          conn_pid: self(),
          pending: %{},
          pending_subscribe: %{},
          subscriptions: %{},
          next_id: 0
        })

      caller_pid = self()

      :sys.replace_state(pid, fn state ->
        %{
          state
          | pending_subscribe:
              Map.put(
                state.pending_subscribe,
                "req-1",
                {"agents.x.events", caller_pid, {caller_pid, make_ref()}}
              )
        }
      end)

      inject_frame(pid, %{
        "jsonrpc" => "2.0",
        "id" => "req-1",
        "result" => %{"subscription" => "sub_abc", "topic" => "agents.x.events"}
      })

      Process.sleep(50)

      state = :sys.get_state(pid)
      assert map_size(state.pending_subscribe) == 0
      assert Map.has_key?(state.subscriptions, "sub_abc")

      GenServer.stop(pid)
    end
  end

  # ── Notification routing ──────────────────────────────────────────────────────

  describe "handle_info/2 — notification routing" do
    test "delivers notification to subscribed process" do
      {:ok, pid} =
        GenServer.start_link(Ws, %Ws{
          conn_pid: self(),
          pending: %{},
          pending_subscribe: %{},
          subscriptions: %{"sub_abc" => [self()]},
          next_id: 0
        })

      inject_frame(pid, %{
        "jsonrpc" => "2.0",
        "method" => "notification",
        "params" => %{
          "subscription" => "sub_abc",
          "event" => "agent.thought",
          "data" => %{"text" => "thinking"}
        }
      })

      assert_receive {:subscription_event, "sub_abc",
                      %{event: "agent.thought", data: %{"text" => "thinking"}}},
                     500

      GenServer.stop(pid)
    end

    test "silently drops notifications for unknown subscription ids" do
      {:ok, pid} =
        GenServer.start_link(Ws, %Ws{
          conn_pid: self(),
          pending: %{},
          pending_subscribe: %{},
          subscriptions: %{},
          next_id: 0
        })

      inject_frame(pid, %{
        "jsonrpc" => "2.0",
        "method" => "notification",
        "params" => %{"subscription" => "ghost", "event" => "x", "data" => nil}
      })

      Process.sleep(50)
      refute_receive {:subscription_event, _, _}, 100

      GenServer.stop(pid)
    end

    test "ignores unknown server-initiated methods" do
      {:ok, pid} =
        GenServer.start_link(Ws, %Ws{
          conn_pid: self(),
          pending: %{},
          pending_subscribe: %{},
          subscriptions: %{},
          next_id: 0
        })

      inject_frame(pid, %{
        "jsonrpc" => "2.0",
        "method" => "session.ready",
        "params" => %{"session_id" => "abc"}
      })

      Process.sleep(50)
      state = :sys.get_state(pid)
      assert map_size(state.subscriptions) == 0

      GenServer.stop(pid)
    end
  end

  # ── Disconnect handling ────────────────────────────────────────────────────────

  describe "handle_info/2 — disconnect" do
    test "clears pending state on disconnect" do
      {:ok, pid} =
        GenServer.start_link(Ws, %Ws{
          conn_pid: self(),
          pending: %{"ws-1" => {self(), make_ref()}},
          pending_subscribe: %{},
          subscriptions: %{"sub_1" => [self()]},
          next_id: 0
        })

      send(pid, :ws_disconnected)
      Process.sleep(50)

      state = :sys.get_state(pid)
      assert map_size(state.pending) == 0
      assert map_size(state.subscriptions) == 0

      GenServer.stop(pid)
    end
  end

  # ── Header helpers ────────────────────────────────────────────────────────────

  describe "to_ws_url / build_headers (via init logic)" do
    test "URL rewriting: https → wss" do
      ws_url =
        "https://gateway.datagrout.ai/servers/test/ws"
        |> String.replace_prefix("https://", "wss://")
        |> String.replace_prefix("http://", "ws://")

      assert ws_url == "wss://gateway.datagrout.ai/servers/test/ws"
    end

    test "URL rewriting: http → ws" do
      ws_url =
        "http://localhost:4000/ws"
        |> String.replace_prefix("https://", "wss://")
        |> String.replace_prefix("http://", "ws://")

      assert ws_url == "ws://localhost:4000/ws"
    end
  end

  # ── Client integration ────────────────────────────────────────────────────────

  describe "DatagroutConduit.Client subscribe/unsubscribe" do
    # These tests exercise handle_call directly with a crafted state (ws_pid: nil)
    # to avoid needing a running GenServer + Mox transport in async mode.

    test "subscribe returns {:error, :not_ws_transport} when ws_pid is nil" do
      state = %DatagroutConduit.Client{ws_pid: nil}

      assert {:reply, {:error, :not_ws_transport}, ^state} =
               DatagroutConduit.Client.handle_call(
                 {:subscribe, "agents.x.events"},
                 {self(), make_ref()},
                 state
               )
    end

    test "unsubscribe returns {:error, :not_ws_transport} when ws_pid is nil" do
      state = %DatagroutConduit.Client{ws_pid: nil}

      assert {:reply, {:error, :not_ws_transport}, ^state} =
               DatagroutConduit.Client.handle_call(
                 {:unsubscribe, "sub_123"},
                 {self(), make_ref()},
                 state
               )
    end
  end

  # ── Ping keepalive ────────────────────────────────────────────────────────

  describe "ping keepalive" do
    test "ping_interval_ms/0 returns 25_000 — Rust parity" do
      assert Ws.ping_interval_ms() == 25_000
    end

    test "default state initialises ping_interval_ms from the module attribute" do
      state = %Ws{conn_pid: self()}
      assert state.ping_interval_ms == 25_000
      assert state.pings_sent == 0
    end

    # Helper: returns a freshly-dead pid.  `WebSockex.send_frame/2` exits
    # with `:noproc` immediately when called against a dead pid — exactly
    # the production crash scenario `safe_send_ping/1` is designed to
    # catch — and it avoids both the 5-second `:gen.call` timeout and the
    # noisy `[error] Agent.Server received unexpected message …` lines
    # that an `Agent.start_link/1` dummy would emit when poked with the
    # `:"$websockex_send"` protocol.
    defp dead_pid do
      pid = spawn(fn -> :ok end)
      ref = Process.monitor(pid)

      receive do
        {:DOWN, ^ref, :process, ^pid, _} -> :ok
      after
        500 -> flunk("spawned process did not exit in time")
      end

      pid
    end

    test ":ping_tick degrades gracefully when the Conn process has died" do
      # WebSockex.send_frame `exit`s when the target Conn is gone;
      # safe_send_ping/1 must catch that so the Ws GenServer doesn't
      # crash alongside it.
      state = %Ws{conn_pid: dead_pid(), ping_interval_ms: 50_000}

      assert {:noreply, new_state} = Ws.handle_info(:ping_tick, state)

      # Send failed, so pings_sent stays at 0 — but the GenServer survived.
      assert new_state.pings_sent == 0
    end

    test ":ping_tick reschedules itself at ping_interval_ms" do
      # Dead conn so the send fails fast; we only care about the
      # reschedule signal here.
      state = %Ws{conn_pid: dead_pid(), ping_interval_ms: 30}

      # First tick fires immediately via direct handle_info call.
      {:noreply, _state} = Ws.handle_info(:ping_tick, state)

      # Production handler scheduled the NEXT tick — verify it lands in
      # this process's mailbox within ~100ms.
      assert_receive :ping_tick, 200
    end

    test "ping_interval_ms = 0 disables rescheduling (no further tick)" do
      state = %Ws{conn_pid: dead_pid(), ping_interval_ms: 0}

      {:noreply, _state} = Ws.handle_info(:ping_tick, state)

      refute_receive :ping_tick, 100
    end

    test "pings_sent/1 returns the live counter through the GenServer API" do
      # Start a Ws GenServer using the test-only init clause that takes a
      # pre-built struct; this avoids the real WS handshake.
      state = %Ws{conn_pid: self(), ping_interval_ms: 0, pings_sent: 7}
      {:ok, pid} = GenServer.start_link(Ws, state)
      assert Ws.pings_sent(pid) == 7
      GenServer.stop(pid)
    end
  end
end
